ptbc5-s2-jon

cysjonathan opened the pull request.

cysjonathan committed:
Add basic clocks

cysjonathan committed:
Add MUI style

cysjonathan committed:
Add analog clock to card

cysjonathan committed:
Add search functionality
Update layout

cysjonathan committed:
Allow cards to be removed

cysjonathan committed:
Add shareable link

cysjonathan committed:
Set date format to reflect 12 instead of 0
Assign toast close responsibility to single close handler

cysjonathan committed:
Cleanup code and unused files

cysjonathan committed:
Add drag-and-drop functionality

cysjonathan committed:
Add night/day style

cysjonathan committed:
Add deployment scripts

cysjonathan committed:
Flush clock cards to top

cysjonathan committed:
Fix issue of city order not preserved

angrylobster commented on src/App.js:
```diff
@@ -1,20 +1,115 @@
+import React, { useEffect, useState } from "react";
+import { CardGrid } from "./components/CardGrid";
+import SelectWithFlag from "./components/SelectWithFlag";
+import _allCities from "./data/cities.json";
```
not sure why there's an underscore before the name here. usually javascript constants you can use either just plain ol' `camelCase` or `CAPITALS_WITH_SNAKE_CASE`

cysjonathan replied:
Got it.
I used `_` to indicate it was a temp holder of the json before I adapted the format into `allCities` below.

angrylobster commented on src/App.js:
```diff
@@ -1,20 +1,115 @@
+  boxShadow: 0,
+}));
+
+const ToastObj = (message, attrs) => ({
```
i would generally avoid using `CapitalCase` for functions. reserve those for either classes or components. maybe this could be called `createToastObj`.

was there a reason for giving each toast a UUID?

cysjonathan replied:
Thanks, updated to `createToastObj`.
On UUID: When the toasts are created in the stack, the child elements need a unique key.
My intention is to clean up completed/closed toast elements, but I have not yet figured a way to do that.

angrylobster replied:
>On UUID: When the toasts are created in the stack, the child elements need a unique key.

makes sense.

>Thanks, updated to createToastObj.
On UUID: When the toasts are created in the stack, the child elements need a unique key.
My intention is to clean up completed/closed toast elements, but I have not yet figured a way to do that.

when i tried creating the toasts with the share button they seem to clear pretty well, after a few seconds.

cysjonathan replied:
The element still lives in the react array though, so it's possible to have a "memory overflow".
https://github.com/jon-rocketacademy-journey/world-clock-bootcamp/blob/dc614ffb9b4fd301208d1124537afcd74c192515/src/App.js#L86
^ keeps adding to the `toasts` array

angrylobster replied:
do you need the `toasts` state? if you avoided keeping a state remembering the toast objects wouldn't that solve your issue of a potential memory overflow? if i'm not wrong the toasts clean up after themselves (i.e. they delete themselves from the DOM)

angrylobster commented on src/App.js:
```diff
@@ -1,20 +1,115 @@
+import ShareIcon from "@mui/icons-material/Share";
+import { useSearchParams } from "react-router-dom";
+
+const Item = styled(Paper)(({ theme }) => ({
```
i think they're deprecating styles from v5 onwards: see https://mui.com/system/styles/basics/

although you can still use styles as a custom styling solution for MUI components, i generally prefer to use in-line styling because it's also easier to debug. see: https://mui.com/system/getting-started/the-sx-prop/

cysjonathan replied:
I see, I had used `styled` copy and pasted from some examples found on MUI docs
https://mui.com/material-ui/react-grid/#css-grid-layout
Will look into the alternatives for future projects.

angrylobster commented on src/App.js:
```diff
@@ -1,20 +1,115 @@
+
+  useEffect(() => {
+    const cityIds = new Set();
+    searchParams.forEach((cityId, _) => cityIds.add(cityId));
```
u can omit the `_` for the index if you're not using it. it's an optional argument since it's the 2nd argument

angrylobster commented on src/App.js:
```diff
@@ -1,20 +1,115 @@
+    const cityIds = new Set();
+    searchParams.forEach((cityId, _) => cityIds.add(cityId));
+    setSelectedCities(allCities.filter((city) => cityIds.has(city.id)));
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
guessing you put this in because you had a warning about not including certain functions varaibles in the dependencies array?

cysjonathan replied:
Correct

angrylobster replied:
do you get stuck in an infinite loop if `searchParams` is a dep?

cysjonathan replied:
I don't know, but I know that I only want the effect to run once, which is why I left it `[]`

angrylobster replied:
in this situation, `searchParams` is derived from `useSearchParams`. the latter is a hook, so it runs on every render. i would say that it makes sense to put `searchParams` as a dependency, because `searchParams` could change, but you wouldn't understand when or why (without reading docs). you should only exclude it as a dependency if it causes you a major bug like infinite loop for re-rendering.

cysjonathan replied:
I just checked, yes it causes an ping-pong effect.
When `searchParams` changes, this effect will `setSelectedCities`, which will call the other effect which tracks `selectedCities`, which in turn will `setSearchParams`.

Ideally we only need to run this once, when the "URL" and its search parameters are provided on page load.
Subsequently, we give up control of the search parameters to the app itself.

angrylobster commented on src/App.js:
```diff
@@ -1,20 +1,115 @@
+  }, []); // run only on first load https://stackoverflow.com/a/55854902
+
+  useEffect(() => {
+    setSearchParams({ c: selectedCities.map((city) => city.id) });
```
hmm, why did you choose to keep track of the cities in the query string of the url? is this for sharing purposes?

cysjonathan replied:
Correct

angrylobster replied:
in that case it might be a good idea to come up with a "[slug](https://medium.com/dailyjs/web-developer-playbook-slug-a6dcbe06c284)" for each one of the cities instead of numbers which are kinda hard to remember. you could create some sort of function to generate the slug by kebab-casing each country and city. 

e.g. `singapore-singapore` or `japan-tokyo` or `american-samoa-taulaga`

angrylobster commented on src/components/AnalogueClock.js:
```diff
@@ -0,0 +1,28 @@
+}
+
+export function AnalogueClock(props) {
+  const { date, timeZone, ...other } = props;
```
not a criticism, but you can also destructure the props directly:

```jsx
export function AnalogueClock({
  date,
  timeZone,
  ...otherProps
}) {
  // component logic here
}
```

cysjonathan replied:
I wanted to be explicit here

angrylobster commented on src/components/CardGridItem.js:
```diff
@@ -0,0 +1,35 @@
+  const { setNodeRef, attributes, listeners, transition, transform } =
+    useSortable({ id: item.id });
+
+  const style = {
```
if the style object is large, or requires a lot of computation, it might be reasonable to use `useMemo` here

cysjonathan replied:
Not sure how to use `useMemo`, would you be able to advise?

angrylobster replied:
`useMemo` (see [docs](https://reactjs.org/docs/hooks-reference.html#usememo)) allows you to cache some value and only recalculate that value when the variables in the dependencies array have changed. 

if you don't use the `useMemo` hook the style object will recalculate on render. `useMemo` is only marginally more effectively if the recalculation is cheap, but it can improve performance significantly if recalculation is expensive.

here, if you wanted to apply it, it would look something like:

```jsx
const style = useMemo(() => ({
  transition,
  transform: CSS.Transform.toString(transform),
}), [transition, transform]);
```

i'm not sure if transition or transform recalculates on every re-render. you will need to test this in order to prevent an infinite loop.

cysjonathan replied:
Thanks!

angrylobster commented on src/components/ClockCard.js:
```diff
@@ -0,0 +1,81 @@
+          hour12: false,
+        })
+      ) +
+        5) %
```
this part here is kinda hard to read. you could either extract some of these into their own constants, or you could use a library like `date-fns` to help convert timezones for you. `date-fns` apis are easy to read and understand

cysjonathan replied:
I moved this to a separate `utils` folder, is this standard practice?
https://github.com/jon-rocketacademy-journey/world-clock-bootcamp/blob/dc614ffb9b4fd301208d1124537afcd74c192515/src/utils/datetime.js

angrylobster replied:
> is this standard practice?

this is a tricky question to answer. it depends on the team. what do i mean? 

well, **in my opinion**, if the utility function is only relevant to that component, then that component should have its own folder and the utility function should be in a `utils.js` file relevant to that component. e.g.

```
| - components
| - - ClockCard
| - - - index.js
| - - - utils.js
```

in the above example, the `index.js` would contain the code for your `ClockCard`  component.

alternatively, if you think / know that the utility function can be reused in other components in the near future, you might have them available in a higher level folder, like what you have done:

```
| - components
| - - ClockCard.js
| - utils
| - - datetime.js
```

some teams might even say that maybe it makes sense to create our own npm library where we have custom functions to handle date time formatting. so instead of stuffing the function into a util file / folder, it may be maintained in a separate repository, then imported via an npm package instead. this use case happens when the devs expect the util function to be reused very frequently, and when there is actual benefit to standardizing across teams how certain util functions work.

```jsx
import { getDateTimeString } from 'jon/utility-functions';
```

the right solution differs a lot from case to case. i would say in this situation the simplest one would suffice - i.e. to group your component and its related files into a single folder.

angrylobster commented on src/components/ClockCard.js:
```diff
@@ -0,0 +1,81 @@
+              hour12: true,
+            })}
+          </Typography>
+          <Typography sx={{ fontSize: "h5.fontSize" }}>
```
might have made sense to fix the height on the title by wrapping it with a div with fixed height. there's an awkward situation where if there are cards with different heights in the same row, they don't display perfectly aligned

<img width="898" alt="image" src="https://user-images.githubusercontent.com/40119924/193591369-70cea13f-8539-4859-a08c-81751f04613d.png">

angrylobster commented on src/components/SelectWithFlag.js:
```diff
@@ -0,0 +1,66 @@
+            width="20"
+            src={`https://flagcdn.com/w20/${option.code.toLowerCase()}.png`}
+            srcSet={`https://flagcdn.com/w40/${option.code.toLowerCase()}.png 2x`}
+            alt=""
```
better to use some kind of identifier for an alt

cysjonathan replied:
True, thanks, I copy pasted the code and forgot about this part.

cysjonathan committed:
Minor nits
Update readme and app title
Slugify city-country
Update drag-and-drop, prevent scroll out of context
Add memo to CardGridItem style
Extract datetime functions into utils folder
Add default popular cities

cysjonathan committed:
Add day/night icon to card

Humanize timezone text, replace underscore with space

cysjonathan committed:
Minor changes
add cities
prevent flash of unstyled card
